### ISSUE TITLE
Thin categories and preorders

### ISSUE THREAD

**Issue by epatters** [epatters/Catlab.jl#84]
On Slack, @kris-brown asks how to do preorders in Catlab. In fact, it would be good to add signatures and syntax systems for preorders and also thin categories to the standard library.

**Comment by epatters** [comment]
Ideally, [thin categories](https://ncatlab.org/nlab/show/thin+category) would be expressed as:

```julia
@theory Category(Ob,Hom) => ThinCategory(Ob,Hom) begin
  (f == g) <= (A::Ob, B::Ob, f::Hom(A,B), g::Hom(A,B))
end
```

However, we cannot yet write this because `@theory` does not exist and `@signature` does not accept axioms (#62). So for now we'd have to write:

```julia
@signature Category(Ob,Hom) => ThinCategory(Ob,Hom) begin
  # (f == g) <= (A::Ob, B::Ob, f::Hom(A,B), g::Hom(A,B))
end
```

**Comment by epatters** [comment]
Preorders should be expressed in the usual terminology, something like:

```julia
@signature Preorder(Ob,Leq) begin
  Ob::TYPE
  Leq(lhs::Ob, rhs::Ob)::TYPE

  reflexive(A::Ob)::Leq(A,A)
  transitive(::Leq(A,B), ::Leq(B,C))::Leq(A,C) <= (A::Ob, B::Ob, C::Ob)
end
```

Note the level shift here: what are considered *axioms* of preorders become *operations* in the generalized algebraic theory.

We should also add syntactic sugar for `\leq` and implement a syntax system `FreePreorder` that normalizes all `Leq`s on the same objects to be equal.

**Comment by jpfairbanks** [comment]
That makes sense to me, but can you explain the part about the level shift? Why do axioms of the preorders become operations in the GAT?

**Comment by epatters** [comment]
Usually, we think of preorders as being a set equipped with a binary relation `<=` (a boolean-valued operation, if you like) satisfying two axioms: reflexivity and transitivity.

But if we want to represent preorders as generalized or essentially algebraic theories, everything has to shift a level downwards. The binary relation `<=` (the "operation") becomes a dependent type (here `Leq`), the reflexivity and transitivity axioms becomes operations, and a new axiom is introduced saying that any two terms of type `Leq(A,B)` are equal. This agrees with the understanding of thin categories as preorders: the composition operation is transitivity and the identity operation is reflexivity. (The associativity and unitality axioms are automatically satisfied due to the thinness axiom, so we don't even have to think about them.)

Is this level shift truly necessary? If so, why? I don't know, actually. I do know, from Porst's paper and [Math.SE](https://math.stackexchange.com/a/2904164), that preorders and posets cannot be axiomatized as algebraic theories. But maybe there are other interesting ways to axiomatize them as GATs.

**Comment by epatters** [comment]
With #62 now resolved, anyone interested could add doctrines for thin categories and preorders.

**Comment by jpfairbanks** [comment]
I think I finally get this. Let me know if there is a misunderstanding. The Catlab issues will be a chronicle of my journey to category theory enlightenment. 

A term constructor can be read as an assertion about all models of the theory, that is, instances of the Doctrine. 

1. To have a term constructor like `reflexive(A::Ob)::Leq(A,A)` you are saying _for all elements A, there exists a term reflexive(A) which implies A<=A_. 

2. And the term constructor `  transitive(::Leq(A,B), ::Leq(B,C))::Leq(A,C) <= (A::Ob, B::Ob, C::Ob)` says _for all objects A,B,C where A<=B and B<= C there exists a term `transitive(A<=B, B<=C)` which implies `A<=C`_. 

So the term constructor `f(args...)::Type{typeargs...} where context` 
are statements like: for all terms satisfying the type constraints in the context, there exists a term `f(args...)` of type `Type{typeargs}`. 

That is how embed mathematical definitions into a GAT by translating a _forall x satisfying P(x), f(x) satisfies Q(X)_ into a term constructor.

**Comment by epatters** [comment]
Exactly. We are using a *term* `Leq` to encode a *relation* `<=`, in which case term constructors play the role of axioms for the relation.

**Comment by jpfairbanks** [comment]
Is this related to how a wiring diagram in a monoidal preorder is a _proof_ of the inequality implied by its domain <= codomain ? In order to construct a new term, your component terms have to satisfy the axioms.

**Comment by epatters** [comment]
Oh, good point. I forgot about that part of Seven Sketches. Yes, that's absolutely the same idea: composite morphisms in form of the wiring diagrams represent proofs of inequality: the inner boxes are the hypothesis inequalities, the outer box is the conclusion inequality, and the diagram itself is the proof.

As part of this issue, we should implement monoidal preorders too!

**Comment by jpfairbanks** [comment]
See #132

**Comment by jpfairbanks** [comment]
I tried to make the objects of the preorder called `Elt` to match the standard notation for preorders where the objects are called elements and the homs are called \le, but that didn't quite work when I went to make the syntax.

**Comment by epatters** [comment]
Thanks for giving this a shot! I commented over at the PR.

**Comment by epatters** [comment]
Resolved by #132 and #148.
